ceph: Add ability to specify compression mode for pools

ShyamsundarR opened the pull request:
**Description of your changes:**
This commit enables configuring a compression_mode for CephPools that takes the values as specified for bluestore OSD `compression_mode` values,
    https://docs.ceph.com/docs/master/rados/configuration/bluestore-config-ref/#inline-compression

**Which issue is resolved by this Pull Request:**
Resolves #5116 

**Checklist:**

- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [x] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
  - [x] Pending EC pool creation unit tests
- [x] Integration tests have been added, if necessary.
  - **NOTE**: I do not think there are any, but would like an opnion if integration tests need to be added for this PR
- [x] Pending release notes updated with breaking and/or notable changes, if necessary.
- [x] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](https://github.com/rook/rook/blob/master/CONTRIBUTING.md#comments)
- [x] Updated example YAMLs with newly added field, for example [pool.yaml](https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/ceph/pool.yaml)

[test ceph]

ShyamsundarR commented on images/image.mk:
```diff
@@ -17,7 +17,9 @@
 
 override GOOS=linux
 
+ifeq ($(origin DOCKERCMD),undefined)
```
:) yes, that one change was for podman. I will separate the commit out.

ShyamsundarR replied:
Fixed.

ShyamsundarR commented on Documentation/ceph-pool-crd.md:
```diff
@@ -83,6 +83,7 @@ When creating an erasure-coded pool, it is highly recommended to create the pool
     > **NOTE**: Neither Rook, nor Ceph, prevent the creation of a cluster where the replicated data (or Erasure Coded chunks) can be written safely. By design, Ceph will delay checking for suitable OSDs until a write request is made and this write can hang if there are not sufficient OSDs to satisfy the request.
 * `deviceClass`: Sets up the CRUSH rule for the pool to distribute data only on the specified device class. If left empty or unspecified, the pool will use the cluster's default CRUSH root, which usually distributes data over all OSDs, regardless of their class.
 * `crushRoot`: The root in the crush map to be used by the pool. If left empty or unspecified, the default root will be used. Creating a crush hierarchy for the OSDs currently requires the Rook toolbox to run the Ceph tools described [here](http://docs.ceph.com/docs/master/rados/operations/crush-map/#modifying-the-crush-map).
+* `compressionMode`: Sets up the pool for inline compression when using a Bluestore OSD. If left unspecified does not setup any compression mode for the pool. Values supported are the same as Bluestore inline compression [modes](https://docs.ceph.com/docs/master/rados/configuration/bluestore-config-ref/#inline-compression), such as `none`, `passive`, `aggressive`, and `force`.
```
Ack! tracked in the checklist.

ShyamsundarR replied:
Fixed

ShyamsundarR commented on pkg/daemon/ceph/client/pool.go:
```diff
@@ -283,6 +292,14 @@ func CreateReplicatedPoolForApp(context *clusterd.Context, namespace, poolName s
 
+	compressionModeProperty := "compression_mode"
+	if pool.CompressionMode != "" {
+		err = SetPoolProperty(context, namespace, poolName, compressionModeProperty, pool.CompressionMode)
```
Yes, I think the minimum version is mimic (although luminous docs. also carry the settings).

ShyamsundarR commented on pkg/apis/ceph.rook.io/v1/types.go:
```diff
@@ -264,6 +264,9 @@ type PoolSpec struct {
 	DeviceClass string `json:"deviceClass"`
 
+	// The inline compression mode in Bluestore OSD to set to (options are: none, passive, aggressive, force)
+	CompressionMode string `json:"compressionMode"`
```
Ok, neat! I did have some questions/confusion here.
- Why does the CRD schema not carry the `DeviceClass` or `CrushRoot` properties, although the types.go has them?
- Should these then also be copied to the CephFilesystem [section](https://github.com/rook/rook/blob/9039c674cef0a90cde22e55318535c4799fa3515/cluster/examples/kubernetes/ceph/common.yaml#L292-L312) and by extension the ObjectStore section etc.? As, we could use a CephFS compressed data pool for similar reasons.
- Ack on the helm chart and integration tests needing the change. Do you see an integration test that needs to be added for this PR? I could not come up with one, hence asking.

ShyamsundarR replied:
Fixed.

But, I still have the questions and possible additions to other CRD kinds like, CephFilesystem pending.

ShyamsundarR replied:
All related changes are done.

I have not added the remaining missing fields with this PR. I can pick that up once this is merged, if it is still pending.

ShyamsundarR commented on pkg/daemon/ceph/client/pool.go:
```diff
@@ -247,6 +248,14 @@ func CreateECPoolForApp(context *clusterd.Context, namespace, poolName, ecProfil
 		}
 	}
 
+	compressionModeProperty := "compression_mode"
```
Fixed

leseb commented on pkg/daemon/ceph/client/pool.go:
```diff
@@ -230,6 +231,26 @@ func givePoolAppTag(context *clusterd.Context, namespace string, poolName string
+	if pool.CompressionMode != "" {
+		err := SetPoolProperty(context, namespace, poolName, compressionModeProperty, pool.CompressionMode)
+		if err != nil {
+			return errors.Wrapf(err, "failed to set property %q on pool %q to %s", compressionModeProperty, poolName, pool.CompressionMode)
```
```suggestion
			return errors.Wrapf(err, "failed to set property %q on pool %q to %q", compressionModeProperty, poolName, pool.CompressionMode)
```

leseb commented on pkg/daemon/ceph/client/pool.go:
```diff
@@ -230,6 +231,26 @@ func givePoolAppTag(context *clusterd.Context, namespace string, poolName string
+	if appName != "" {
+		err := givePoolAppTag(context, namespace, poolName, appName)
+		if err != nil {
+			return err
```
please wrap the error with a meaningful error message

leseb commented on pkg/operator/ceph/pool/validate.go:
```diff
@@ -87,5 +87,15 @@ func ValidatePoolSpec(context *clusterd.Context, namespace string, p *cephv1.Poo
+		case "none", "passive", "aggressive", "force":
+			break
+		default:
+			return errors.Errorf("unrecognized compression mode %s", p.CompressionMode)
```
```suggestion
			return errors.Errorf("unrecognized compression mode %q", p.CompressionMode)
```

leseb commented on tests/framework/installer/ceph_manifests.go:
```diff
@@ -506,6 +506,13 @@ spec:
                   type: integer
                   minimum: 0
                   maximum: 9
+            compressionMode:
```
Please also update `GetBlockPoolDef()` so that the CI can test compression too.

ShyamsundarR commented:
@leseb all comments addressed. Thanks!

mykaul commented on cluster/examples/kubernetes/ceph/filesystem.yaml:
```diff
@@ -26,6 +26,8 @@ spec:
         # gives a hint (%) to Ceph in terms of expected consumption of the total cluster capacity of a given pool
         # for more info: https://docs.ceph.com/docs/master/rados/operations/placement-groups/#specifying-expected-pool-size
         #targetSizeRatio: .5
+      #compressionMode: none | passive | aggressive | force
```
Is there any drawback (performance?) to having the default compression mode being 'passive'? 
@liewegas  ?
(The paranoid in me would suspect a non-cooperating client might try to DoS us with in-compressible data, but that's a bit far fetched, no? there are other ways it can probably cause us to work hard).

ShyamsundarR replied:
> How about uncommenting it and having the default be `none`? Then the setting is more visible and the comment can contain the options.

Done, I prefer them visible as well, otherwise it is lost in the comment noise at times.

As I changed these to be visible, I also changes the *-test.yaml files to add the required `none` values, such that these are tested during the integration tests.

ShyamsundarR commented on cluster/examples/kubernetes/ceph/filesystem.yaml:
```diff
@@ -26,6 +26,8 @@ spec:
         # for more info: https://docs.ceph.com/docs/master/rados/operations/placement-groups/#specifying-expected-pool-size
         #targetSizeRatio: .5
+      #compressionMode: none | passive | aggressive | force
+      # For reference: https://docs.ceph.com/docs/nautilus/rados/configuration/bluestore-config-ref/#inline-compression
```
Fixed with the latest update.

ShyamsundarR commented on cluster/examples/kubernetes/ceph/filesystem-ec.yaml:
```diff
@@ -21,6 +21,9 @@ spec:
     - erasureCoded:
         dataChunks: 2
         codingChunks: 1
+      # Inline compression mode for the data pool
```
Left one instance of the doc link in each of Block, CephFS, RGW instances and removed the rest.

ShyamsundarR commented:
Previous instance of the PR failed integration tests, as all metadata pools took the default empty string value for the newly added `compressionMode` which was not a valid value. This is now corrected.

Further, updated the file `upgrade-from-v1.2-crds.yaml` to add the changes to the CRDs.
